optimum/neuron/modeling_seq2seq.py

```diff
@@ -0,0 +1,912 @@
+                )
+
+            if return_dict_in_generate and output_scores:
+                beam_indices = tuple((beam_indices[beam_idx[i]] + (beam_idx[i],) for i in range(len(beam_indices))))
+
+            # increase cur_len
+            cur_len = cur_len + 1
+
+            # stop when each sentence is finished, or if we exceed the maximum length
+            stop_criterion_1 = beam_scorer.is_done
+            if isinstance(stopping_criteria, list):
+                if len(stopping_criteria) == 1:
+                    stopping_criteria = stopping_criteria[0]
+
+            # Cases that can be handled in XLA without requiring
+            # non-padded input_ids
+            if isinstance(stopping_criteria, MaxLengthCriteria):
+                stop_criterion_2 = cur_len >= stopping_criteria.max_length
+            elif isinstance(stopping_criteria, MaxTimeCriteria):
+                stop_criterion_2 = stopping_criteria(input_ids, scores)
+            else:
+                # Other cases will be handled on CPU
+                batch_size, _ = input_ids.shape
+                input_ids_cpu = input_ids.to("cpu")
+                mask = torch.cat(
+                    [torch.ones(batch_size, cur_len), torch.zeros(batch_size, input_ids.shape[1] - cur_len)], dim=1
+                ).bool()
+                input_ids_cpu = torch.masked_select(input_ids_cpu, mask).reshape((batch_size, cur_len))
+                scores_cpu = scores.to("cpu") if torch.is_tensor(scores) else scores
+                stop_criterion_2 = stopping_criteria(input_ids_cpu, scores_cpu)
+
+            if stop_criterion_1 or stop_criterion_2:
+                if not synced_gpus:
+                    break
+
+        sequence_outputs = beam_scorer.finalize(
+            input_ids.to("cpu"),
+            beam_scores.to("cpu"),
+            next_tokens.to("cpu"),
+            next_indices.to("cpu"),
+            pad_token_id=pad_token_id,
+            eos_token_id=eos_token_id,
+            max_length=stopping_criteria.max_length,
+            beam_indices=beam_indices,
+        )
+
+        for k, v in sequence_outputs.items():
+            if type(v) == torch.Tensor:
+                sequence_outputs[k] = sequence_outputs[k].to(input_ids.device)
+
+        return sequence_outputs["sequences"]
+
+    def greedy_search(
+        self,
+        input_ids: torch.LongTensor,
+        logits_processor: Optional["LogitsProcessorList"] = None,
+        stopping_criteria: Optional["StoppingCriteriaList"] = None,
+        max_length: Optional[int] = None,
+        pad_token_id: Optional[int] = None,
+        eos_token_id: Optional[Union[int, List[int]]] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        output_scores: Optional[bool] = None,
+        return_dict_in_generate: Optional[bool] = None,
+        seq_length: Optional[int] = int,
+        streamer: Optional["BaseStreamer"] = None,
```

Comment:
You never stream the tokens. If that is not supported, then we should remove it from the generate parameters.

Reply:
I'm not so familiar with how the streamer is used, isn't it the same as how we did in the `NeuronGenerationMixin`, shall we remove it as well there?

https://github.com/huggingface/optimum-neuron/blob/50ec86f387f0220c2f890749d7b776169c34c7da/optimum/neuron/generation/utils.py#L1072

Reply:
If a streamer is provided you are supposed to call streamer.put() every time a token has been generated. When you're done you also need to call streamer.end().
Anyway, I don't think we need to support it as we typically deploy inference endpoints using TGI now, not using generate at all.

Reply:
Ok, I will remove all streamer then